completion_aggregator/tasks.py

```diff
@@ -0,0 +1,188 @@
+        total_possible = 0.0
+        last_modified = OLD_DATETIME
+        for child in self._get_children(block):
+            (earned, possible, modified) = self.update_for_block(child, changed_blocks)
+            total_earned += earned
+            total_possible += possible
+            last_modified = max(last_modified, modified)
+        if self._aggregator_needs_update(block, last_modified):
+            Aggregator.objects.submit_completion(
+                user=self.user,
+                course_key=self.course_key,
+                block_key=block,
+                aggregation_name=block.block_type,
+                earned=total_earned,
+                possible=total_possible,
+                last_modified=last_modified,
+            )
+        return (total_earned, total_possible, last_modified)
+
+    def update_for_excluded(self):
+        """
+        Return a sentinel empty completion value for excluded blocks.
+        """
+        return (0.0, 0.0, OLD_DATETIME)
+
+    def update_for_completable(self, block, changed_blocks):  # pylint: disable=unused-argument
+        """
+        Return the block completion value for a given completable block.
+        """
+        completion = self.block_completions.get(block)
+        if completion:
+            earned = completion.completion
+            modified = completion.modified
+        else:
+            earned = 0.0
+            modified = OLD_DATETIME
+        return (earned, 1.0, modified)
+
+    def _aggregator_needs_update(self, block, modified):
+        """
+        Return True if the given aggregator block needs to be updated.
+
+        This method assumes that the block has already been determined to be an aggregator.
+        """
+        agg = self.aggregators.get(block)
+        return is_registered_aggregator(block) and getattr(agg, 'last_modified', OLD_DATETIME) < modified
+
+    # Dependency isolation methods
+
+    def init_course_block_key(self, modulestore, course_key):
+        """
+        Return a UsageKey for the root course block.
+        """
+        # pragma: no-cover
+        return modulestore.make_course_usage_key(course_key)
+
+    def init_course_blocks(self, user, course_block_key):
+        """
+        Return a BlockStructure representing the course.
+
+        Blocks must have the following attributes:
+
+            .location
+            .block_type
+        """
+        # pragma: no-cover
+        from lms.djangoapps.course_blocks.api import get_course_blocks  # pylint: disable=import-error
+        return get_course_blocks(user, course_block_key)
+
+    def _get_block_completions(self):
+        """
+        Return the list of BlockCompletions.
+
+        Each must have the following attributes:
+
+            .block_key (UsageKey)
+            .modified (datetime)
+            .completion
+        """
+        from completion.models import BlockCompletion  # pylint: disable=import-error
+        return BlockCompletion.objects.filter(
```

Comment:
There's a classmethod on the `BlockCompletion` model now that will do this for you and also give you the dictionary you construct in `__init__()`. https://github.com/edx/completion/blob/master/completion/models.py#L188

Reply:
Spiffy!  I'll put it to good use.

Reply:
Looking closer at this, I'd like to include this work in #6 instead.  Because it changes the abstraction barriers, introducing this here would make merging #6 significantly more painful.

Reply:
👍 Gotchya